src/pip/_internal/req/__init__.py

```diff
@@ -60,39 +68,89 @@ def install_given_reqs(
+        )
+    )
 
     with indent_log():
-        for req_name, requirement in to_install.items():
-            if requirement.should_reinstall:
-                logger.info('Attempting uninstall: %s', req_name)
-                with indent_log():
-                    uninstalled_pathset = requirement.uninstall(
-                        auto_confirm=True
-                    )
-            else:
-                uninstalled_pathset = None
-
-            try:
-                requirement.install(
-                    install_options,
-                    global_options,
-                    root=root,
-                    home=home,
-                    prefix=prefix,
-                    warn_script_location=warn_script_location,
-                    use_user_site=use_user_site,
-                    pycompile=pycompile,
-                )
-            except Exception:
-                # if install did not succeed, rollback previous uninstall
-                if uninstalled_pathset and not requirement.install_succeeded:
-                    uninstalled_pathset.rollback()
-                raise
-            else:
-                if uninstalled_pathset and requirement.install_succeeded:
-                    uninstalled_pathset.commit()
-
-            installed.append(InstallationResult(req_name))
+        # first try to install in parallel
+        installed_pool = map_multiprocess_ordered(
+            partial(_single_install, install_args, in_subprocess=True),
+            requirements)
+
+        # check the results from the parallel installation,
+        # and fill-in missing installations or raise exception
+        for installed_req, requested_req in zip(installed_pool, requirements):
+            # if the requirement was not installed by the parallel pool,
+            # install serially here
+            if installed_req is None:
+                installed_req = _single_install(
+                    install_args, requested_req, in_subprocess=False)
+
+            if isinstance(installed_req, BaseException):
+                # Raise an exception if we caught one
+                # during the parallel installation
+                raise installed_req
+            elif isinstance(installed_req, InstallationResult):
+                installed.append(installed_req)
 
     return installed
+
+
+def _single_install(
+        install_args,           # type: _InstallArgs
+        requirement,            # type: InstallRequirement
+        in_subprocess=False,    # type: bool
+):
+    # type: (...) -> Union[None, InstallationResult, BaseException]
+    """
+    Install a single requirement, returns InstallationResult
+    (to be called per requirement, either in parallel or serially)
+    """
+
+    # if we are running inside a subprocess,
+    # then only clean wheel installation is supported
+    if (in_subprocess and
+            (requirement.should_reinstall or not requirement.is_wheel)):
+        return None
```

Comment:
I’m wondering whether it’d be easier to prevent spawning a process at all for this branch.

Reply:
This is a good point, I'll refactor it. It will not be very clean based on your input that OrderedDict is not to be trusted, so I will make do with index list to preserve package install order.